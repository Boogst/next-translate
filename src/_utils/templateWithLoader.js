import { getNormalExportRegex } from './hasExportName'

export default function templateWithLoader(
  rawCode,
  { page, typescript, loader, hasLoader } = {}
) {
  let modifiedCode = rawCode

  if (hasLoader) {
    modifiedCode = modifiedCode.replace(getNormalExportRegex(loader), (v) =>
      v.replace('export', 'const _getStaticProps = ')
    )
  }

  let template = `
    import __i18nConfig from '${process.cwd() + '/i18n'}'
    import __loadNamespaces from 'next-translate/loadNamespaces'
    ###__CURRENT_CODE_HERE__###
    export async function ${loader}(ctx) {
        ${hasLoader ? `let res = _${loader}(ctx)` : ''}
        ${hasLoader ? `if(typeof res.then === 'function') res = await res` : ''}
        return {
          ${hasLoader ? '...res,' : ''}
          props: {
            ${hasLoader ? '...(res.props || {}),' : ''}
            ...(await __loadNamespaces({
              ...ctx,
              pathname: '${page}',
              ...__i18nConfig,
              defaultLoader: (l, n) => import(\`${
                process.cwd() + '/locales/${l}/${n}'
              }\`)
                .then(m => m.default)
            }))
          }
        }
    }
  `

  if (typescript) template = template.replace(/\n/g, '\n// @ts-ignore\n')

  return template.replace('###__CURRENT_CODE_HERE__###', `\n${modifiedCode}\n`)
}
