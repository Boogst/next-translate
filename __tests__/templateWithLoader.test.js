import templateWithLoader from '../src/_utils/templateWithLoader'

const clean = (t) => t.replace(/( |\n|\t|\u00A0)/g, '')

const tests = [
  {
    describe: 'exporting getStaticProps is a function',
    code: `
      export default function Page() {
        return <div>Hello world</div>
      }

      export function getStaticProps(){
        return { {props: {}} }
      }
  `,
    cases: [
      {
        i18nFile: '/i18n.js',
        page: '/index',
        typescript: false,
        loader: 'getStaticProps',
        hasLoader: true,
        prefix: '..',
        expected: `
        import __i18nConfig from '../i18n.js'
        import getPageNamespaces from 'next-translate/_utils/getPageNamespaces'

        export default function Page() {
          return <div>Hello world</div>
        }
  
        const _getStaticProps = function getStaticProps(){
          return { {props: {}} }
        }

        export async function getStaticProps(ctx) {
           const __lang = ctx.locale
           let res = _getStaticProps(ctx)
           if(typeof res.then === 'function') res = await res

           const defaultLoader = (l, n) => import(\`../locales/\${l}/\${n}.json\`).then(m => m.default)
           const loader = __i18nConfig.loadLocaleFrom || defaultLoader
           const namespaces = await getPageNamespaces(__i18nConfig, '/', ctx)
           const pns = await Promise.all(namespaces.map(ns => loader(__lang, ns)))
           const __namespaces = namespaces.reduce((obj, ns, i) => { obj[ns] = pns[i]; return obj }, {})

           return { 
              ...res, 
              props: {
                ...(res.props || {}),
                __namespaces,
                __lang,
              }
            }
        }
    `,
      },
    ],
  },
  {
    describe: 'without loader (getStaticProps as default)',
    code: `
      export default function Page() {
        return <div>Hello world</div>
      }
  `,
    cases: [
      {
        i18nFile: '/i18n.json',
        page: '/index',
        typescript: false,
        loader: 'getStaticProps',
        hasLoader: false,
        prefix: '..',
        expected: `
        import __i18nConfig from '../i18n.json'
        import getPageNamespaces from 'next-translate/_utils/getPageNamespaces'

        export default function Page() {
          return <div>Hello world</div>
        }

        export async function getStaticProps(ctx) {
           const __lang = ctx.locale

           const defaultLoader = (l, n) => import(\`../locales/\${l}/\${n}.json\`).then(m => m.default)
           const loader = __i18nConfig.loadLocaleFrom || defaultLoader
           const namespaces = await getPageNamespaces(__i18nConfig, '/', ctx)
           const pns = await Promise.all(namespaces.map(ns => loader(__lang, ns)))
           const __namespaces = namespaces.reduce((obj, ns, i) => { obj[ns] = pns[i]; return obj }, {})

           return { 
              props: {
                __namespaces,
                __lang,
              }
            }
        }
    `,
      },
    ],
  },
]

describe('templateWithLoader', () => {
  tests.forEach((d) => {
    describe(d.describe, () => {
      d.cases.forEach(({ expected, debug, ...options }) => {
        const fn = debug ? test.only : test
        const testname = Object.entries(options).map(([k, v]) => `${k}: ${v}`)
        fn(testname.join(' | '), () => {
          expect(clean(templateWithLoader(d.code, options))).toEqual(
            clean(expected)
          )
        })
      })
    })
  })
})
