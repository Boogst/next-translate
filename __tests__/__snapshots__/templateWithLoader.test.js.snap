// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`templateWithLoader exporting as a arrow function - without loader page: /blog/[post]/[...catchall] | loader: getServerSideProps | hasLoader: false 1`] = `
"import __i18nConfig from \\"/i18n\\";
import __loadNamespaces from \\"next-translate/loadNamespaces\\";

export default () => <div>Hello world</div>;

export async function getServerSideProps(ctx) {
  return {
    props: {
      ...(await __loadNamespaces({
        ...ctx,
        pathname: \\"/blog/[post]/[...catchall]\\",
        ...__i18nConfig,
        defaultLoader: (l, n) =>
          import(\`/Users/aroca/i18n-next-static/locales/\${l}/\${n}\`).then(
            (m) => m.default
          ),
      })),
    },
  };
}
"
`;

exports[`templateWithLoader exporting as a arrow function - without loader page: /index | loader: getStaticProps | hasLoader: false 1`] = `
"import __i18nConfig from \\"/i18n\\";
import __loadNamespaces from \\"next-translate/loadNamespaces\\";

export default () => <div>Hello world</div>;

export async function getStaticProps(ctx) {
  return {
    props: {
      ...(await __loadNamespaces({
        ...ctx,
        pathname: \\"/index\\",
        ...__i18nConfig,
        defaultLoader: (l, n) =>
          import(\`/Users/aroca/i18n-next-static/locales/\${l}/\${n}\`).then(
            (m) => m.default
          ),
      })),
    },
  };
}
"
`;

exports[`templateWithLoader exporting as a class - with loader page: /index | loader: getStaticProps | hasLoader: true 1`] = `
"import __i18nConfig from \\"/i18n\\";
import __loadNamespaces from \\"next-translate/loadNamespaces\\";

import React from \\"react\\";

export default class Page extends React.Component {
  render() {
    return <div>Hello world</div>;
  }
}

const _getStaticProps = function getStaticProps() {
  return { props: {} };
};

export async function getStaticProps(ctx) {
  let res = _getStaticProps(ctx);
  if (typeof res.then === \\"function\\") res = await res;
  return {
    ...res,
    props: {
      ...(res.props || {}),
      ...(await __loadNamespaces({
        ...ctx,
        pathname: \\"/index\\",
        ...__i18nConfig,
        defaultLoader: (l, n) =>
          import(\`/Users/aroca/i18n-next-static/locales/\${l}/\${n}\`).then(
            (m) => m.default
          ),
      })),
    },
  };
}
"
`;

exports[`templateWithLoader exporting as a class in a diferent line - without loader page: /blog/[post]/[...catchall] | loader: getServerSideProps | hasLoader: false 1`] = `
"import __i18nConfig from \\"/i18n\\";
import __loadNamespaces from \\"next-translate/loadNamespaces\\";

import React from \\"react\\";

class Page extends React.Component {
  render() {
    return <div>Hello world</div>;
  }
}

const myPage = Page;

export default myPage;

export async function getServerSideProps(ctx) {
  return {
    props: {
      ...(await __loadNamespaces({
        ...ctx,
        pathname: \\"/blog/[post]/[...catchall]\\",
        ...__i18nConfig,
        defaultLoader: (l, n) =>
          import(\`/Users/aroca/i18n-next-static/locales/\${l}/\${n}\`).then(
            (m) => m.default
          ),
      })),
    },
  };
}
"
`;

exports[`templateWithLoader exporting as a class in a diferent line - without loader page: /index | loader: getStaticProps | hasLoader: false 1`] = `
"import __i18nConfig from \\"/i18n\\";
import __loadNamespaces from \\"next-translate/loadNamespaces\\";

import React from \\"react\\";

class Page extends React.Component {
  render() {
    return <div>Hello world</div>;
  }
}

const myPage = Page;

export default myPage;

export async function getStaticProps(ctx) {
  return {
    props: {
      ...(await __loadNamespaces({
        ...ctx,
        pathname: \\"/index\\",
        ...__i18nConfig,
        defaultLoader: (l, n) =>
          import(\`/Users/aroca/i18n-next-static/locales/\${l}/\${n}\`).then(
            (m) => m.default
          ),
      })),
    },
  };
}
"
`;

exports[`templateWithLoader exporting as a function - with loader page: /index | loader: getStaticProps | hasLoader: true 1`] = `
"import __i18nConfig from \\"/i18n\\";
import __loadNamespaces from \\"next-translate/loadNamespaces\\";

export default function Page() {
  return <div>Hello world</div>;
}

export function getStaticPaths() {
  return {};
}

const _getStaticProps = function getStaticProps() {
  return { props: {} };
};

export async function getStaticProps(ctx) {
  let res = _getStaticProps(ctx);
  if (typeof res.then === \\"function\\") res = await res;
  return {
    ...res,
    props: {
      ...(res.props || {}),
      ...(await __loadNamespaces({
        ...ctx,
        pathname: \\"/index\\",
        ...__i18nConfig,
        defaultLoader: (l, n) =>
          import(\`/Users/aroca/i18n-next-static/locales/\${l}/\${n}\`).then(
            (m) => m.default
          ),
      })),
    },
  };
}
"
`;

exports[`templateWithLoader exporting as a function - without loader page: /blog/[post]/[...catchall] | loader: getServerSideProps | hasLoader: false 1`] = `
"import __i18nConfig from \\"/i18n\\";
import __loadNamespaces from \\"next-translate/loadNamespaces\\";

export default function Page() {
  return <div>Hello world</div>;
}

export async function getServerSideProps(ctx) {
  return {
    props: {
      ...(await __loadNamespaces({
        ...ctx,
        pathname: \\"/blog/[post]/[...catchall]\\",
        ...__i18nConfig,
        defaultLoader: (l, n) =>
          import(\`/Users/aroca/i18n-next-static/locales/\${l}/\${n}\`).then(
            (m) => m.default
          ),
      })),
    },
  };
}
"
`;

exports[`templateWithLoader exporting as a function - without loader page: /index | loader: getStaticProps | hasLoader: false 1`] = `
"import __i18nConfig from \\"/i18n\\";
import __loadNamespaces from \\"next-translate/loadNamespaces\\";

export default function Page() {
  return <div>Hello world</div>;
}

export async function getStaticProps(ctx) {
  return {
    props: {
      ...(await __loadNamespaces({
        ...ctx,
        pathname: \\"/index\\",
        ...__i18nConfig,
        defaultLoader: (l, n) =>
          import(\`/Users/aroca/i18n-next-static/locales/\${l}/\${n}\`).then(
            (m) => m.default
          ),
      })),
    },
  };
}
"
`;

exports[`templateWithLoader exporting as a function in diferent line - with loader page: /index | loader: getServerSideProps | hasLoader: true 1`] = `
"import __i18nConfig from \\"/i18n\\";
import __loadNamespaces from \\"next-translate/loadNamespaces\\";

function Page() {
  return <div>Hello world</div>;
}

function _getServerSideProps() {
  return { props: {} };
}

// export { getServerSideProps };
export default Page;

export async function getServerSideProps(ctx) {
  let res = _getServerSideProps(ctx);
  if (typeof res.then === \\"function\\") res = await res;
  return {
    ...res,
    props: {
      ...(res.props || {}),
      ...(await __loadNamespaces({
        ...ctx,
        pathname: \\"/index\\",
        ...__i18nConfig,
        defaultLoader: (l, n) =>
          import(\`/Users/aroca/i18n-next-static/locales/\${l}/\${n}\`).then(
            (m) => m.default
          ),
      })),
    },
  };
}
"
`;

exports[`templateWithLoader exporting as a function in diferent line - without loader page: /blog/[post]/[...catchall] | loader: getServerSideProps | hasLoader: false 1`] = `
"import __i18nConfig from \\"/i18n\\";
import __loadNamespaces from \\"next-translate/loadNamespaces\\";

function Page() {
  return <div>Hello world</div>;
}

export default Page;

export async function getServerSideProps(ctx) {
  return {
    props: {
      ...(await __loadNamespaces({
        ...ctx,
        pathname: \\"/blog/[post]/[...catchall]\\",
        ...__i18nConfig,
        defaultLoader: (l, n) =>
          import(\`/Users/aroca/i18n-next-static/locales/\${l}/\${n}\`).then(
            (m) => m.default
          ),
      })),
    },
  };
}
"
`;

exports[`templateWithLoader exporting as a function in diferent line - without loader page: /index | loader: getStaticProps | hasLoader: false 1`] = `
"import __i18nConfig from \\"/i18n\\";
import __loadNamespaces from \\"next-translate/loadNamespaces\\";

function Page() {
  return <div>Hello world</div>;
}

export default Page;

export async function getStaticProps(ctx) {
  return {
    props: {
      ...(await __loadNamespaces({
        ...ctx,
        pathname: \\"/index\\",
        ...__i18nConfig,
        defaultLoader: (l, n) =>
          import(\`/Users/aroca/i18n-next-static/locales/\${l}/\${n}\`).then(
            (m) => m.default
          ),
      })),
    },
  };
}
"
`;

exports[`templateWithLoader loader as a arrow function page: /index | loader: getStaticProps | hasLoader: true 1`] = `
"import __i18nConfig from \\"/i18n\\";
import __loadNamespaces from \\"next-translate/loadNamespaces\\";

export default function Page() {
  return <div>Hello world</div>;
}

const _getStaticProps = () => {
  props: {
  }
};
//export { getStaticProps };

export async function getStaticProps(ctx) {
  let res = _getStaticProps(ctx);
  if (typeof res.then === \\"function\\") res = await res;
  return {
    ...res,
    props: {
      ...(res.props || {}),
      ...(await __loadNamespaces({
        ...ctx,
        pathname: \\"/index\\",
        ...__i18nConfig,
        defaultLoader: (l, n) =>
          import(\`/Users/aroca/i18n-next-static/locales/\${l}/\${n}\`).then(
            (m) => m.default
          ),
      })),
    },
  };
}
"
`;

exports[`templateWithLoader loader imported to another place page: /index | loader: getStaticProps | hasLoader: true 1`] = `
"import __i18nConfig from \\"/i18n\\";
import __loadNamespaces from \\"next-translate/loadNamespaces\\";

import getStaticProps from \\"somewhere/getStaticProps\\";
import _getStaticPaths from \\"somewhere/getStaticPaths\\";

const config = {};

export default function Page() {
  return <div>Hello world</div>;
}

export { config, getStaticPaths };

export async function getStaticProps(ctx) {
  let res = _getStaticProps(ctx);
  if (typeof res.then === \\"function\\") res = await res;
  return {
    ...res,
    props: {
      ...(res.props || {}),
      ...(await __loadNamespaces({
        ...ctx,
        pathname: \\"/index\\",
        ...__i18nConfig,
        defaultLoader: (l, n) =>
          import(\`/Users/aroca/i18n-next-static/locales/\${l}/\${n}\`).then(
            (m) => m.default
          ),
      })),
    },
  };
}
"
`;

exports[`templateWithLoader loader with named import to another place + rename page: /index | loader: getStaticProps | hasLoader: true 1`] = `
"import __i18nConfig from \\"/i18n\\";
import __loadNamespaces from \\"next-translate/loadNamespaces\\";

import { getStaticPropsA as _getStaticProps } from \\"somewhere/getStaticProps\\";
import { getStaticProps as getStaticPaths } from \\"somewhere/getStaticPaths\\";

const config = {};

export default function Page() {
  return <div>Hello world</div>;
}

export { config, getStaticPaths };

export async function getStaticProps(ctx) {
  let res = _getStaticProps(ctx);
  if (typeof res.then === \\"function\\") res = await res;
  return {
    ...res,
    props: {
      ...(res.props || {}),
      ...(await __loadNamespaces({
        ...ctx,
        pathname: \\"/index\\",
        ...__i18nConfig,
        defaultLoader: (l, n) =>
          import(\`/Users/aroca/i18n-next-static/locales/\${l}/\${n}\`).then(
            (m) => m.default
          ),
      })),
    },
  };
}
"
`;

exports[`templateWithLoader loader with named import to another place page: /index | loader: getStaticProps | hasLoader: true 1`] = `
"import __i18nConfig from \\"/i18n\\";
import __loadNamespaces from \\"next-translate/loadNamespaces\\";

import { getStaticProps as _getStaticProps } from \\"somewhere/getStaticProps\\";
import { getStaticPaths } from \\"somewhere/getStaticPaths\\";

const config = {};

export default function Page() {
  return <div>Hello world</div>;
}

export { config, getStaticPaths };

export async function getStaticProps(ctx) {
  let res = _getStaticProps(ctx);
  if (typeof res.then === \\"function\\") res = await res;
  return {
    ...res,
    props: {
      ...(res.props || {}),
      ...(await __loadNamespaces({
        ...ctx,
        pathname: \\"/index\\",
        ...__i18nConfig,
        defaultLoader: (l, n) =>
          import(\`/Users/aroca/i18n-next-static/locales/\${l}/\${n}\`).then(
            (m) => m.default
          ),
      })),
    },
  };
}
"
`;

exports[`templateWithLoader loader with one named import to another place page: /index | loader: getStaticProps | hasLoader: true 1`] = `
"import __i18nConfig from \\"/i18n\\";
import __loadNamespaces from \\"next-translate/loadNamespaces\\";

import {
  getStaticPaths,
  getStaticProps as _getStaticProps,
  config,
} from \\"somewhere/getStaticProps\\";

export default function Page() {
  const test = \\"getStaticProps\\";
  console.log(\\"This should log getStaticProps without any modification\\");
  return <div>Hello getStaticProps</div>;
}

export { config, getStaticPaths };

export async function getStaticProps(ctx) {
  let res = _getStaticProps(ctx);
  if (typeof res.then === \\"function\\") res = await res;
  return {
    ...res,
    props: {
      ...(res.props || {}),
      ...(await __loadNamespaces({
        ...ctx,
        pathname: \\"/index\\",
        ...__i18nConfig,
        defaultLoader: (l, n) =>
          import(\`/Users/aroca/i18n-next-static/locales/\${l}/\${n}\`).then(
            (m) => m.default
          ),
      })),
    },
  };
}
"
`;
